cip-caseygibson-dynamicsaturation.md

```diff
@@ -0,0 +1,88 @@
+
+While unlikely, the system does not take into account external influences (eg. political) that could harm the system. Shelly has been active for less then 2 years and we are already seeing the risks on a small scale. As real world adoption continues, there could be situations in the next decade that could jeopardise the decentralisation of Cardano. One example could be that a political party could run a ISPO (Initial Stake Pool Offering) on a mass scale. Seeing as Cardano has the potential to be used for voting (even at a political level), the integrity of Cardano could be questioned if political parties controlled stake pools with large shares of total stake. 
+
+## Motivation
+
+The pledge used in a pool was meant to show how serious a SPO is and how much "skin in the game" they had. The idea was that if they have more pledge, they have more to lose if something goes wrong. This seems to have lost it's meaning as SPOs that already have a dominating position have created many sub-pools. Some have split up their pledge evenly, some have next to no pledge and have gained high amounts of stake through influential means. This has not only reduced the security of Cardano, but has also lost the meaning of having pledge in the pool. 
+
+For example, a SPO might have a pledge of 30,000 ADA across 10 pools, while another SPO might have 1,000,000 pledge but only has 1 pool with a small amount of active stake. Seeing as the SPO with 1M pledge has more overall, they have more to lose and should be trusted more. Since there is no technical advantage to having a high pledge, the meaning and purpose of a pledge is redundant.
+
+To make the pledge a meaningful metric that is fair to all SPOs and aligns with the core values of Cardano, the pledge should be used to calculate the saturation point of a pool. This will mean that SPOs, no matter how many pools they operate, will have a maximum saturation based on their total pledge. For example, if a pool operates a single pool and wanted to open up another pool with the same amount of stake, they will have to assign the equal amount of pledge against that new pool. If a pool wishes to up their saturation point, they will need to assign a higher amount of pledge.
+
+## Specification
+
+For this to be able to work, there firstly needs to be an upper limit and a lower limit. The K parameter can still be used as the upper saturation limit of a single pool. As in, if a SPO has enough pledge assigned to a single pool, that pool will be able to run at the maximum saturation point of K. The lower limit is in place to safe guard small SPOs and allow them to grow.
+
+An example of how Dynamic Saturation would be calculated:
+
+500,000 ADA Pledged = Saturation point at 100% of K
```

Comment:
We will need a mechanism to determine these pledge thresholds. For example, why does 500,000 ADA pledge, instead of 600,000 ADA, deserve 100% saturation point.

Reply:
It's not smart in general to limit saturation, dynamic impact of pledge, relative to total delegated stake would be much more beneficial, and provide a level playing field...

Reply:
> We will need a mechanism to determine these pledge thresholds. For example, why does 500,000 ADA pledge, instead of 600,000 ADA, deserve 100% saturation point.

The example I gave was based on the fact that the values would be adjustable in the same way all the other params were chosen and could be changed based on network consensus.

Reply:
> It's not smart in general to limit saturation, dynamic impact of pledge, relative to total delegated stake would be much more beneficial, and provide a level playing field...

For the purpose of the discussion, can you please explain why "it's not good to limit saturation"? Technically the k param already limits saturation, so I'm not sure what your point is.

The idea is to shift the saturation to the pledge instead of on the pool.

Reply:
Here is my idea of formulating the thresholds as a function of `k` and two additional parameters, `e` (entry threshold) and `l` (leverage):

```
orig_sat = total_supply / k
new_sat = orig_sat * max(e, min(1 / k, pledge / orig_sat * l))
```

We can use function `new_sat` to calculate the saturation point for a pool given its pledge, so we don't need to define these thresholds manually in the protocol.The parameter `e` is to, like Casey mentioned, safe guard small SPOs so they can at least have a reasonable chance to generate a block in a certain period of time. The parameter `l` (leverage) could be a value the community define, e.g. 20, which means the saturation point will be 20x the amount of pledge. 

@mark-stopka Not limiting the saturation point means we will discard `k` completely, which will be a much bigger change than the scope of this proposal I think.

Reply:
Thanks for the formula @cffls. Based on my initial analysis it looks pretty promising.

I wrote up your formula as Javascript just to make sure we're clear on your thinking and to see what it would look like.

I played around with `e` and `l` as I thought they were a bit aggressive. In the CIP proposal I picked 500,000 pledge as it seems like a reasonable amount based on what I've seen from pool data. So I've tried to compress it to the lower end so that max saturation is reasonably achievable.

It would be good to see arguments for shifting the range either way.

```
lovelace = 1000000;

function calc_sat(pledge){
    k = 500;
    e = 0.2;
    l = 125;
    total_supply = 33719282563 * lovelace;
    orig_sat = total_supply / k;
    new_sat = orig_sat * Math.max(e, min(1 / k, pledge / orig_sat * l));
    final_sat = max(new_sat, orig_sat);
    console.log(`pledge: ${pledge / lovelace}, sat: ${final_sat / lovelace}`);
}

function max(val1, val2){
    if(val1 < val2){
        return val1;
    }
    return val2;
}

function min(val1, val2){
    if(val1 > val2){
        return val1;
    }
    return val2;
}

calc_sat(50000 * lovelace);
calc_sat(100000 * lovelace);
calc_sat(150000 * lovelace);
calc_sat(250000 * lovelace);
calc_sat(500000 * lovelace);
calc_sat(750000 * lovelace);
calc_sat(1000000 * lovelace);
calc_sat(2000000 * lovelace);

```

results:
```
[Log] pledge: 50000, sat: 13487713.0252
[Log] pledge: 100000, sat: 13487713.0252
[Log] pledge: 150000, sat: 18750000
[Log] pledge: 250000, sat: 31250000
[Log] pledge: 500000, sat: 62500000
[Log] pledge: 750000, sat: 67438565.126
[Log] pledge: 1000000, sat: 67438565.126
[Log] pledge: 2000000, sat: 67438565.126
```

Overall it does do basically what I had in mind.

Reply:
Thanks Casey @CyberCyclone  for testing my idea and sharing the result here. I didn't think much about the values 
 for `e` and `l`. Yours (`e=0.2` and `l=125`) seems more reasonable than mine. Another good thing about formulating the saturation point is that we can tune parameters (`e` and `l`) based on the price of ADA and `k`, and eventually have community to vote and decide their values.

Reply:
as a small pool, reaching saturation at 13m stake vs 64m seems more reasonable to me and targeting 50k pledge also seems like a realistic target this would encourage me to increase my pledge to avoid declining rewards for my delegates - do I understand this correctly?

Reply:
These numbers are just used for discussion purpose and nothing is finalized. In this context, increasing the pledge means increasing the saturation threshold for your pool. If your pool's pledge is at 50k, then the pool will be saturated at 13.49M ADA, meaning that any more delegation to the pool will result in decreasing reward. In this case, you are correct in a sense that you would need to increase the pledge above 50k in order to receive more delegations beyond 13.49M ADA without hurting the ROI of your pool.

Reply:
@CyberCyclone Casey, if this formula looks good to you, would you mind including it in your proposal? I can also create a pull request to your clone, and then you can merge it from your clone and resubmit this pull request. I am fine with either way.

Reply:
Thanks for the reminder @cffls Jerry. I'll try and look into adding it in the next few days. I'm working on another CIP so once that's ready I'll come back to this one.

Reply:
@cffls I've finally gotten around to updating the CIP. I've attached the code to the bottom of the CIP.